### PR TITLE
WIP: Add postpublish hook and built-in pre and post hooks

### PR DESCRIPTION
**Pull request by gregorylegarec**
To do:

* [x] Fix tests in related commit (where postpublish calls are added)
* [x] Write documentation
* [ ] Rewrite mattermost post hook in nodeJS and handle multiple instances update in the same message
* [ ] Add documentation on how ro generate [this line](https://github.com/cozy/cozy-collect/blob/master/.travis.yml#L33)

### CODE REVIEW

File: lib/manual.js [hunk 1]
```diff
@@ -91,19 +91,20 @@ async function manualPublish ({
...
   prompt.message = colorize.bold('Confirmation:')
   prompt.delimiter = ' '
-  prompt.get(promptProperties, function (err, received) {
+  prompt.get(promptProperties, async function (err, received) {
+    let publication
     console.log()
     if (err) throw new Error(colorize.red(`prompt: ${err}`))
     if (received.confirm.match(/^y(es)?$/i)) {
-      publish(publishOptions, finishCallback)
+      publication = await publish(publishOptions)
```
Comment (CPatchane) [inline lib/manual.js]: Since we don't use it I think we can remove it

**Comment by y-lohse** [comment]
[Déploiement automatique de drive, photos et contacts](https://trello.com/c/32FZtgFn/1031-d%C3%A9ploiement-automatique-de-drive-photos-et-contacts)

**Comment by y-lohse** [comment]
I've bullet-style documented the workflow here : https://gist.github.com/y-lohse/8dfe1323c5c6180ff2ac4238dd955db9

File: .eslintrc.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+    "extends": ["cozy-app"],
+    "rules": {
+      "no-console": 0
```
Comment (kosssi) [inline .eslintrc.json]: Si ça désactive eslint pour les `console.log`, je serai plus pour laisser la configuration de base de cozy-app et d'ajouter des `// eslint-disable-next-line no-console` pour expliciter que c'est pas un `console.log` de dev oublié

Comment (y-lohse) [reply .eslintrc.json]: Comme c'est un script node qui utilise la console pour communiquer avec l'utilisateur, je pense que c'est assez approprié. Il faudrait ajouter *beaucoup* de `// eslint-disable-next-line no-console` au point ou je pense que ca ne serait plus très explicite (quand une ligne comme ca est répétée trop souvent, on fini par l'ignorer).

**Comment by y-lohse** [comment]
I've added a general description of the workflow, please correct me if I'm wrong.

I'm pretty sure we're still lacking a few things — such as automatically creating the release branch with a predefined text, process details, etc — but I think at this point the best way forward is to merge this PR and give it a shot with cozy-drive or some other app. This should highlight what's missing and we can come back here and complete things.

What do you guys think ?

File: README.md [hunk 3]
```diff
@@ -111,14 +111,81 @@ This function must return an object containing the same options given in paramet
...
+```
+cozy-app-publish --prepublish rundeck
+```
+
+Deploys the app on rundeck. This hook requires several variables to be set as environment variables:
+
+- `RUNDECK_TOKEN`:  the token to use to interact with the Rundeck API
+- `TARGETS_DEV`: a comma-separated list of instances to deploy the app on, on the `dev` channel.
+- `TARGETS_BETA`: a comma-separated list of instances to deploy the app on, on the `beta` channel.
+- `TARGETS_STABLE`: a comma-separated list of instances to deploy the app on, on the `stable` channel.
```
Comment (ptbrowne) [inline README.md]: Would be cool if those environment variables were checked before the hook is run to fail faster. Each hook would need a `validate` function of some kind.

File: README.md [hunk 4]
```diff
@@ -111,14 +111,81 @@ This function must return an object containing the same options given in paramet
...
+#### `--verbose`
 
 To print more logs when using tool (useful for debug).
+
+### Recommended workflow
+
+#### Day to day
+
+- Development is done on feature branches that are merged into `master`,  once they are complete.
+- Every time someone commits on `master`, a new archive is created an uploaded on Downcloud and it is then deployed on a test instance with Rundeck.
```
Comment (ptbrowne) [inline README.md]: is created *and* uploaded

File: .eslintrc.json (hunk 2 above)
Comment (ptbrowne) [reply .eslintrc.json]: Petite pensée je ne sais pas ce qu'elle vaut. Peut être que dans ce cas il vaudrait mieux avoir une fonction `output` ou `messageUser` ou `log` avec 1 seul `// eslint-disable-next-line no-console` à l'intérieur.

File: lib/hooks/post/rundeck.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  if (version.includes('-dev.')) return DEV
+  if (version.includes('-beta.')) return BETA
+  if (version.match(/\d+\.\d+\.\d+/)) return STABLE
+  throw new Error('Unrecognized version channel')
+}
+
+const getInstanceTLD = instance => {
+  return instance
+    .split('.')
+    .splice(1)
```
Comment (ptbrowne) [inline lib/hooks/post/rundeck.js]: `slice` is clearer here.

File: lib/hooks/post/rundeck.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+const STABLE = 'stable'
+
+const getChannel = version => {
+  if (version.includes('-dev.')) return DEV
+  if (version.includes('-beta.')) return BETA
+  if (version.match(/\d+\.\d+\.\d+/)) return STABLE
+  throw new Error('Unrecognized version channel')
+}
+
+const getInstanceTLD = instance => {
```
Comment (ptbrowne) [inline lib/hooks/post/rundeck.js]: TLD is confusing as in "toto.cozy.works", it means "works". Here I would use `domain`.

File: lib/hooks/pre/downcloud.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env node
+
+import fs from 'fs-extra'
+import tar from 'tar'
+import { spawn } from 'child_process'
+
+const BUILD_FOLDER = './build/'
+const COZY_URL = 'downcloud.cozycloud.cc'
```
Comment (ptbrowne) [inline lib/hooks/pre/downcloud.js]: URL only ?

File: lib/hooks/pre/downcloud.js [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+const launchCmd = (cmd, params, options) => {
```
Comment (ptbrowne) [inline lib/hooks/pre/downcloud.js]: `spawnPromise` would be clearer IMHO.

File: lib/hooks/pre/downcloud.js (hunk 7 above)
Comment (y-lohse) [reply lib/hooks/pre/downcloud.js]: Do you mean there should be a protocol or something? The URL gets completed in various places in various ways, and this appears to be the common part.

File: lib/hooks/pre/downcloud.js (hunk 8 above)
Comment (y-lohse) [reply lib/hooks/pre/downcloud.js]: Meh, can't say I agree tbh. `spawn` is very node-oriented, but I think `launchCmd ` can be understood by everyone. And I don't think the function name should reflect that it returns a promise.

**Comment by y-lohse** [comment]
Thanks for the review, especially given the mamoth-sized diff. Fixes incoming!

**Comment by gregorylegarec** [comment]
As the number of commits is increasing and a lot of commits are `fix` commits, it could be nice to rebase/merge them into the commit they are fixing.

File: README.md [hunk 9]
```diff
@@ -173,3 +173,19 @@ Use this options to provide a specific space name of the registry to publish the
...
+
+#### Release workflow
+
+- A new branch is created from the current state of `master`. Let's say we want to deploy version `1.0.0` of the app.
```
Comment (gregorylegarec) [inline README.md]: We may clarfiy here by being more explicit:
> A release branch is created from the current state of `master`. `master` minor version is increased, i.e `master` is now at `1.1.0`.

File: README.md [hunk 10]
```diff
@@ -173,3 +173,19 @@ Use this options to provide a specific space name of the registry to publish the
...
+- The only type of commits allowed on this branch are bug fixes.
```
Comment (gregorylegarec) [inline README.md]: s/this branch/this release branch/

File: README.md [hunk 11]
```diff
@@ -173,3 +173,19 @@ Use this options to provide a specific space name of the registry to publish the
...
+- Every commit is tagged with a prerelease version number, eg. `1.0.0-beta.1`, `1.0.0-beta.2`, etc…
```
Comment (gregorylegarec) [inline README.md]: Actually, not every commit is tagged. A tag is generated every time we want to publish a new beta. A new beta may contains several fixes and several commits.

**Comment by gregorylegarec** [comment]
@y-lohse Thanks for the workflow documentation !

**Comment by CPatchane** [comment]
Thanks and nice job @gregorylegarec @y-lohse 🎉

**Comment by gregorylegarec** [comment]
@CPatchane Maybe I am wrong but it seems that the mattermost script has not been rewritten in nodeJS and it still publish only on `recette.cozy.works`, and don't take into account `TARGETS_DEV`, `TARGETS_BETA`, `TARGET_PROD`, etc.

The way to generate the line https://github.com/cozy/cozy-collect/blob/master/.travis.yml#L34 has not been documented.

Also, fix commits have not been merged and squashed to keep a clean tree and clean commits.

**Comment by y-lohse** [comment]
Oh, my bad !

I thought we had agreed on doing the bash to node conversion at some pointer later, but I should have asked. And you're correct about it not taking the `TARGETS_` into account. I'll fill an issue so we don't forget.

@ptbrowne gave me the link to the travis documentation about generating the cryptic line. As mentioned in [this earlier comment](https://github.com/cozy/cozy-app-publish/pull/35#issuecomment-405615598), I'll update the docs again after a succesfull setup, including docs about this line.

I did rebase most of the commits though. I think the remaining ones are general fixes that aren't related to one other commit in particular? https://github.com/cozy/cozy-app-publish/pull/35/commits/0294b03c74820afabef6d4e3f20fb87f1729b910 could have been merged with the rundeck commit through, my bad again.

Apologies if the merge was premature. I agree there's still work left to do, but I figured that everything was stable and that testing the hooks out for real would be a good next step at this point.

**Pull request merged**